app/src/main/java/org/mozilla/fenix/tabstray/syncedtabs/SyncedTabsComposables.kt

```diff
@@ -0,0 +1,321 @@
+                            }
+                        )
+                ) {
+                    Icon(
+                        painter = painterResource(id = R.drawable.ic_sign_in),
+                        contentDescription = null,
+                        tint = FirefoxTheme.colors.textInverted,
+                    )
+                    Spacer(modifier = Modifier.width(8.dp))
+                    Text(
+                        text = stringResource(id = R.string.synced_tabs_sign_in_button),
+                        color = FirefoxTheme.colors.textInverted,
+                        fontSize = 14.sp,
+                        fontFamily = FontFamily(Font(R.font.metropolis_semibold)),
+                        maxLines = 2,
+                        modifier = Modifier.align(Alignment.CenterVertically)
+                    )
+                }
+            }
+        }
+    }
+}
+
+@Composable
+@Preview
+private fun SyncedTabsDeviceErrorPreview() {
+    FirefoxTheme {
+        Box(Modifier.background(FirefoxTheme.colors.layer2)) {
+            SyncedTabsErrorItem(SyncedTabsListItem.Error(R.string.synced_tabs_no_tabs))
+        }
+    }
+}
+
+@Composable
+fun SyncedTabsNoTabsItem() {
+    SecondaryText(
+        text = stringResource(id = R.string.synced_tabs_no_open_tabs),
+        fontSize = 12.sp,
+        maxLines = 1,
+        modifier = Modifier
+            .padding(PaddingValues(16.dp, 7.dp))
+            .fillMaxWidth()
+    )
+}
+
+@Composable
+@Preview
+private fun SyncedTabsNoTabsItemPreview() {
+    FirefoxTheme {
+        Box(Modifier.background(FirefoxTheme.colors.layer2)) {
+            SyncedTabsNoTabsItem()
+        }
+    }
+}
+
+@Composable
+@Preview
+private fun SyncedTabsListPreview() {
+    FirefoxTheme {
+        Box(Modifier.background(FirefoxTheme.colors.layer2)) {
+            SyncedTabsList(
+                getFakeSyncedTabList(),
+                object : SyncedTabsView.Listener {
+                    override fun onRefresh() = Unit
+
+                    override fun onTabClicked(tab: mozilla.components.browser.storage.sync.Tab) {
+                        println("Tab clicked")
+                    }
+                }
+            )
+        }
+    }
+}
+
+/**
+ * The various types of list items that can be found in a [SyncedTabsList].
+ */
+sealed class SyncedTabsListItem {
+
+    /**
+     * A device header for displaying a synced device.
+     */
+    data class Device(val displayName: String, val id: String) : SyncedTabsListItem()
+
+    /**
+     * A tab that was synced.
+     */
+    data class Tab(val tab: SyncTab) : SyncedTabsListItem()
+
+    /**
+     * A placeholder for a device that has no tabs synced.
+     */
+    data class NoTabs(val deviceId: String) : SyncedTabsListItem()
+
+    /**
+     * A message displayed if an error was encountered.
+     */
+    data class Error(
+        val descriptionResId: Int,
```

Comment:
Rather than storing the descriptionResId, maybe it would be nicer to store the ErrorType and then use the extension functions to derive the stringResId in our composable functions. Just a thought if you are open to entertaining it.

Reply:
This makes sense